### PR TITLE
attr interface for SymbolicNodes

### PR DESCRIPTION
**Pull request by vchuravy**
Follow up on #35.

I went with the Nullable abstraction, because we don't have `haskey` and it encapsulates what is happening underneath. 

@pluskid Where would the right place to document this be?

### CODE REVIEW

File: src/symbolic-node.jl [hunk 1]
```diff
@@ -133,9 +134,9 @@ function get_attr(self :: SymbolicNode, key :: Symbol)
   ref_success = Ref{Cint}(-1)
   @mxcall(:MXSymbolGetAttr, (MX_handle, Cstring, Ref{Cstring}, Ref{Cint}), self, key_s, ref_out, ref_success)
   if ref_success[] == 1
-    return bytestring(ref_out[])
+    return Nullable(bytestring(ref_out[]))
```
Comment (pluskid) [inline src/symbolic-node.jl]: Can you use `Nullable{AbstractString}` explicitly?

File: src/symbolic-node.jl [hunk 2]
```diff
@@ -133,9 +134,9 @@ function get_attr(self :: SymbolicNode, key :: Symbol)
...
   else
-    throw(KeyError(key))
+    return Nullable()
```
Comment (pluskid) [inline src/symbolic-node.jl]: Can you use `Nullable{AbstractString}` explicitly?

File: src/symbolic-node.jl [hunk 3]
```diff
@@ -164,10 +165,14 @@ end
    Create a symbolic variable with the given name. This is typically used as a placeholder.
    For example, the data node, acting as the starting point of a network architecture.
 =#
-function Variable(name :: Union{Base.Symbol, AbstractString})
+function Variable(name :: Union{Base.Symbol, AbstractString}; attrs = Dict{Symbol, ASCIIString}())
```
Comment (pluskid) [inline src/symbolic-node.jl]: Why `ASCIIString` is used instead of `AbstractString`?

**Comment by pluskid** [comment]
@vchuravy Thanks! I had some inline comments.

BTW: the right place to document what?

**Comment by vchuravy** [comment]
The right place to address the attr keyword for SymbolicNodes and Variables.

**Comment by pluskid** [comment]
The place for `Variable` doc should go [here](https://github.com/dmlc/MXNet.jl/blob/master/src/symbolic-node.jl#L162), and the doc for general operators could be appended [here](https://github.com/dmlc/MXNet.jl/blob/master/src/symbolic-node.jl#L492).

File: src/symbolic-node.jl [hunk 4]
```diff
@@ -159,15 +160,19 @@ function set_attr(self :: SymbolicNode, key :: Symbol, value :: AbstractString)
...
 
 #=doc
-.. function:: Variable(name :: Union{Base.Symbol, AbstractString})
+.. function:: Variable(name :: Union{Symbol, AbstractString})
 
    Create a symbolic variable with the given name. This is typically used as a placeholder.
    For example, the data node, acting as the starting point of a network architecture.
 =#
-function Variable(name :: Union{Base.Symbol, AbstractString})
+function Variable(name :: Union{Symbol, AbstractString}; attrs = Dict{Symbol, AbstractString}())
```
Comment (pluskid) [inline src/symbolic-node.jl]: Julia types not being covariant bite me a lot. This interface might be somehow inconvenient to use, as if the user write
```
attrs = Dict(:foo => "bar", :bar => "baz")
```
they will get a `Dict{Symbol,ASCIIString}` (or `UTF8String` depending on whether they used unicode), yet that is not a subtype of `Dict{Symbol,AbstractString}`, and the call will fail with a signature mismatch. 

A workaround that I'm currently using is in the signature, use general `attrs::Dict = Dict()` type, and then in the function body, do a explicit call to
```
attrs = convert(Dict{Symbol,AbstractString}, attrs)
```
attempt to get the right behavior and fail gracefully with meaningful error message. Can you modify it to be of this way? Thanks!

File: src/symbolic-node.jl [hunk 5]
```diff
@@ -521,6 +527,12 @@ function _define_atomic_symbol_creator(hdr :: MX_handle; gen_docs=false)
       if k == :name; continue; end
       if isa(v, SymbolicNode)
         symbol_kws[k] = v
+      elseif k == :attrs
+        if isa(v, Dict)
+          attrs = v
```
Comment (pluskid) [inline src/symbolic-node.jl]: Can you also add
```
v = convert(Dict{Symbol,AbstractString}, v)
```
here?

**Comment by vchuravy** [comment]
@pluskid done and also see the new test to ensure that it is the correct behaviour.

**Comment by pluskid** [comment]
Thanks a lot!

**Pull request merged**